workflows/cloudwatch_logs_big_query/SPECIFICATION.md

```diff
@@ -0,0 +1,125 @@
+
+Queries are jobs. You can start a query with `StartQuery`, but it immediately returns the `queryId`. You must poll a query using `GetQueryResults` until the query has finished. For the purpose of this example, a query has "finished" when `GetQueryResults` has returned a status of one of "Complete", "Failed", "Cancelled", "Timeout", or "Unknown".
+
+`StartQuery` responds with an error if the query's start or end date occurs out of bounds of the log group creation date. The error message starts with "Query's end date and time".
+
+Start the query and wait for it to "finish". Store the `results`. If the count of the results is less than the configured LIMIT, return the results. If the the results are greater than or equal to the limit, go to [Recursive queries](#recursive-queries).
+
+---
+
+### Recursive queries
+
+If the result count from the previous step is 10000 (or the configured LIMIT), it is very likely that there are more results. The example should do a binary search of the remaining logs. To do this, get the date of the last (earliest or latest depending on sort order) log. Use that date as the start date of a new date range, while the end date can remain the same.
+
+Split that date range in half, resulting in two new date ranges. Call your query function twice; once for each new date range.
+
+Concatenate the results of the first query with the results of the two new queries.
+
+The following pseudocode illustrates this.
+
+```pseudocode
+func big_query(date_range):
+  query_results = get_query_results(date_range)
+
+  if query_results.length < LIMIT
+    return query_results
+  else
+    date_range = [query_results.end, date_range.end]
+    d1, d2 = split(date_range)
+    return concat(query_results, big_query(d1), big_query(d2))
+```
+
+## Output
+
+To illustrate the search, log the date ranges for each query made and the number of logs that were found.
+
+Example:
+
+```
+Query date range: 2023-12-21T14:58:16.000Z to 2023-12-21T15:03:15.994Z. Found 10000 logs.
+Query date range: 2023-12-21T15:01:15.995Z to 2023-12-21T15:03:15.994Z. Found 10000 logs.
+Query date range: 2023-12-21T14:59:15.995Z to 2023-12-21T15:01:15.994Z. Found 10000 logs.
+Query date range: 2023-12-21T15:02:15.995Z to 2023-12-21T15:02:45.994Z. Found 5000 logs.
+Query date range: 2023-12-21T15:02:45.995Z to 2023-12-21T15:03:15.994Z. Found 5000 logs.
+Query date range: 2023-12-21T15:00:45.995Z to 2023-12-21T15:01:15.994Z. Found 5000 logs.
+Query date range: 2023-12-21T15:00:15.995Z to 2023-12-21T15:00:45.994Z. Found 5000 logs.
+Queries finished in 12.372 seconds.
+Total logs found: 50000
+```
```

Comment:
I would move or copy this to the central README where the user will be following steps and wanting to validate the output they see is correct.

Reply:
I don't think that's the appropriate place for this. This is example output that we (the team) can use as a guide, it's not descriptive.